Bug fixes and E2E tests for PVC storage

ckadner opened the pull request:
### Motivation

Address PVC follow-up work items outlined in #337 for PVC storage introduced in #267

### Modifications

Code changes:
- Sort PVC mounts on serving runtime specs to avoid unstable repeated
  runtime rollouts as Kubernetes treat two otherwise identical deployment
  specs as different if the same set of volume mounts are in different order
- Don't add non-existent PVCs from predictor/ISVC when `allowAnyPVC` is
  enabled as this would cause all serving pods for that runtime to stay in
  Pending state with unbound (pending) volumes
- Tolerate missing `storage-config` secret when `allowAnyPVC` is enabled
- Lint: fix `"io/ioutil"` deprecations

FVT changes:
- Add Storage test suite
- Add helper methods to add PVC to storage-config during FVT
- Allow for additional time in WaitForReadyDeployStatus but allow early abort on success
- Check if pod still running before gRPC/REST requests, reconnect if necessary
- Only choose "Ready" runtime pod for port-forwards
- Include ISVC tests in Predictor test suite to ensure "serial" execution of TLS tests

Resolves #337

ckadner commented on Makefile:
```diff
@@ -46,8 +46,10 @@ test:
+# TODO: reinstate all FVT suites
 fvt:
-	ginkgo -v -p -progress --fail-fast fvt/predictor fvt/scaleToZero --timeout=40m
+	@#ginkgo -v -p --fail-fast fvt/predictor fvt/scaleToZero fvt/storage --timeout=40m
```
temporarily remove predictor tests and scaleToZero test to isolate PVC test cases

ckadner commented on config/runtimes/ovms-1.x.yaml:
```diff
@@ -57,3 +57,6 @@ spec:
     modelLoadingTimeoutMillis: 90000
+
+  # TODO: re-enable OpenVino runtime
+  disabled: true
```
temporarily disabled OpenVino runtime to isolate PVC test cases

ckadner commented on config/runtimes/triton-2.x.yaml:
```diff
@@ -93,3 +93,6 @@ spec:
     modelLoadingTimeoutMillis: 90000
+
+  # TODO: re-enable Triton runtime
+  disabled: true
```
temporarily disabled Triton runtime to isolate PVC test cases

ckadner commented:
Somehow this PR broke. Maybe one too many force pushes. It complains about code that does not exist on this fork:

```
> Run ./scripts/develop.sh make fmt
which: no controller-gen in (/root/go/bin/:/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/kubebuilder/bin:)
./scripts/fmt.sh || (echo "Linter failed: $?"; git status; exit 1)
golangci-lint............................................................Failed
- hook id: golangci-lint
- exit code: 1

controllers/suite_test.go:230:53: undeclared name: `ioutil` (typecheck)
		if payloadProcessingTestConfigFileContents, err = ioutil.ReadFile(testConfigFile); err != nil {
		                                                  ^

prettier.................................................................Passed
Pre-commit linter failed, but it may have automatically formatted your files.
Check your changed files and/or manually fix the errors above then stage and commit.
Linter failed: 1
HEAD detached at pull/340/merge
nothing to commit, working tree clean
make: *** [Makefile:108: fmt] Error 1
Error: Process completed with exit code 2.
```

https://github.com/ckadner/modelmesh-serving/blob/pvc_fvt/controllers/suite_test.go#L230

ckadner commented:
> Somehow this PR broke. Maybe one too many force pushes. It complains about code that does not exist on this fork:
> 
> ```
> > Run ./scripts/develop.sh make fmt
> which: no controller-gen in (/root/go/bin/:/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/kubebuilder/bin:)
> ./scripts/fmt.sh || (echo "Linter failed: $?"; git status; exit 1)
> golangci-lint............................................................Failed
> - hook id: golangci-lint
> - exit code: 1
> 
> controllers/suite_test.go:230:53: undeclared name: `ioutil` (typecheck)
> 		if payloadProcessingTestConfigFileContents, err = ioutil.ReadFile(testConfigFile); err != nil {
> 		                                                  ^
> 
> prettier.................................................................Passed
> Pre-commit linter failed, but it may have automatically formatted your files.
> Check your changed files and/or manually fix the errors above then stage and commit.
> Linter failed: 1
> HEAD detached at pull/340/merge
> nothing to commit, working tree clean
> make: *** [Makefile:108: fmt] Error 1
> Error: Process completed with exit code 2.
> ```
> 
> https://github.com/ckadner/modelmesh-serving/blob/pvc_fvt/controllers/suite_test.go#L230

Oh, it's an inconsistency since #339 was merged, my deprecation-linted imports would cause a compile error with some of the code added by Rob Geada. I will try to rebase this PR.

ckadner commented:
@njhill -- I think the PVC support fixes are ready for your review now.

@tjohnson31415
When you have some time, could you review the [FVT changes](https://github.com/kserve/modelmesh-serving/pull/340/files#diff-7e6b0ecf385a7b9657f26b65b99769c75344a972a07a4583b34e5279b33cae82)?

@rafvasq
If you can spare a few minutes, please help take a look at the [updated storage config docs](https://github.com/kserve/modelmesh-serving/pull/340/files#diff-a120db8f77d13164b6d8b82caa72baf994a3f7184b0f39e3cef6702132b2baa2)

Thank you guys and sorry for the large PR but there was no good way to separate the code fixes from the new FVT cases.

rafvasq commented on docs/predictors/setup-storage.md:
```diff
@@ -95,3 +97,239 @@ azureKey: |
+
+Let's update (or create) the `model-serving-config` ConfigMap.
+
+Note, if you already have a `model-serving-config` ConfigMap, you might want to retain the existing config overrides. You can check you current config flags by running:
```
```suggestion
Note, if you already have a `model-serving-config` ConfigMap, you might want to retain the existing config overrides. You can check your current configuration flags by running:
```

rafvasq commented on docs/predictors/setup-storage.md:
````diff
@@ -95,3 +97,239 @@ azureKey: |
+kubectl get cm "model-serving-config" -o jsonpath="{.data['config\.yaml']}"`
+```
+
+The minimal `model-serving-config` for our example only needs the two settings `allowAnyPVC` and `restProxy` enabled true:
````
```suggestion
The minimal `model-serving-config` for our example only requires the settings `allowAnyPVC` and `restProxy` to be enabled:
```

rafvasq commented on docs/predictors/setup-storage.md:
````diff
@@ -95,3 +97,239 @@ azureKey: |
+EOF
+```
+
+After a few seconds the new InferenceService `sklearn-pvc-example` should be ready:
````
```suggestion
After a few seconds, the new InferenceService `sklearn-pvc-example` should be ready:
```

rafvasq commented on docs/predictors/setup-storage.md:
```diff
@@ -95,3 +97,239 @@ azureKey: |
+
+### 3. (b) Enable `allowAnyPVC` in the `model-serving-config` ConfigMap
+
+As an alternative to preconfiguring all _allowed_ PVCs in the `storage-config` secret, you can set the `allowAnyPVC` configuration flag to `true`. With `allowAnyPVC` enabled, users can deploy Predictors or InferenceService with models stored on _any_ PVC in the model serving namespace.
```
```suggestion
As an alternative to preconfiguring all _allowed_ PVCs in the `storage-config` secret, you can set the `allowAnyPVC` configuration flag to `true`. With `allowAnyPVC` enabled, users can deploy Predictors or InferenceServices with models stored on _any_ PVC in the model serving namespace.
```

rafvasq commented on docs/predictors/setup-storage.md:
```diff
@@ -95,3 +97,239 @@ azureKey: |
+
+## Deploy a model stored on a Persistent Volume Claim
+
+Models can be stored on Kubernetes Persistent Volumes.
```
```suggestion
Models can be stored on [Kubernetes Persistent Volumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/).
```

rafvasq commented on docs/predictors/setup-storage.md:
```diff
@@ -1,14 +1,16 @@
+You will need access to an S3-compatible object storage, for example [MinIO](https://github.com/minio/minio). To configure access to the object storage, use the `storage-config` secret.
 
-## Deploy a model from your own object storage
+Alternatively, models can be stored on a Kubernetes Persistent Volume. Persistent Volume Claims can either be pre-configured in the `storage-config` secret, or, the `allowAnyPVC` config flag can be enabled, so that any PVC can be mounted dynamically at the time a predictor or inference service is deployed.
```
```suggestion
Alternatively, models can be stored on a Kubernetes Persistent Volume. Persistent Volume Claims can either be pre-configured in the `storage-config` secret, or, the `allowAnyPVC` configuration flag can be enabled, so that any PVC can be mounted dynamically at the time a predictor or inference service is deployed.
```

rafvasq reviewed (commented):
The docs look good to me Christian. Just left a few nit/typo catches, nothing big :)

tjohnson31415 commented on fvt/storage/storage_suite_test.go:
```diff
@@ -0,0 +1,135 @@
+	RunSpecs(t, "Storage Suite")
+}
+
+func createFVTClient() {
```
This looks to be copied as-is from the predictor suite. It could be moved to `fvt/utils.go` to have the code shared (and probably should be renamed to something like `initializeFVTSuite()`.

ckadner replied:
it's kind of the same code in all 3 suites. I can refactor it

ckadner replied:
I did the refactoring for all 3 suites

ckadner commented on docs/release-process.md:
```diff
@@ -243,7 +243,7 @@ release blog.
+- [v0.7.0](https://kserve.github.io/website/0.7/blog/articles/2021-10-11-KServe-0.7-release)
 - [v0.8.0](https://kserve.github.io/website/0.8/blog/articles/2022-02-18-KServe-0.8-release/#modelmesh-updates)
 - [v0.9.0](https://kserve.github.io/website/0.9/blog/articles/2022-07-21-KServe-0.9-release/#inferenceservice-api-for-modelmesh)
 - [v0.10.0 draft](https://github.com/kserve/website/pull/227#discussion_r1098917277)
```
Thanks @rafvasq -- I updated the link and added some of the PRs to illustrate the actual authoring process

tjohnson31415 commented on fvt/storage/storage_test.go:
```diff
@@ -0,0 +1,189 @@
+					ExpectSuccessfulInference_sklearnMnistSvm(isvcName)
+				})
+
+				BeforeAll(func() {
```
NIT: to match the order of execution of the code better (and the other test suites), I think `BeforeAll` and `AfterAll` should be above the first `It` in this `Describe`.

ckadner replied:
sure, will move it

ckadner replied:
done

tjohnson31415 commented on fvt/storage/storage_test.go:
```diff
@@ -0,0 +1,189 @@
+			//   the old pod (without the PVC) but the old pod gets kept around
+			//   instead of the new because the ISVC is still on there -- even
+			//   though its failing
+			// - allowAnyPVC needs rest-proxy enabled (not sure why)
```
Hmm, that seems quirky. Hopefully that is something we can fix easily.

njhill replied:
Yes this sounds pretty odd, it should be pretty much unrelated. Do you have any more detail of the actual error/symptoms?

ckadner replied:
Oh, this part: `allowAnyPVC needs rest-proxy enabled (not sure why)`? 

Sorry that was from an early misreading of Chin's code. I will remove it.

```go

	// collect PVCs from Predictors when the global flag 'allowAnyPVC' is enabled
	predictorPVCsMap := make(map[string]struct{})
	if cfg.AllowAnyPVC {
		restProxyEnabled := cfg.RESTProxy.Enabled

		// use a predicate function to extract the PVCs from Predictors in the registry
		f := func(p *api.Predictor) bool {
			if runtimeSupportsPredictor(rt, p, restProxyEnabled, req.Name) &&
				p.Spec.Storage != nil &&
```

ckadner replied:
some of the other thoughts/obeservations/assumptions in my comment are no longer true since we fixed the unstable deployments

ckadner replied:
I cleared that up -- apologies this confused rambling were my working notes while trying to figure out what was causing the deployment instabilities

tjohnson31415 commented on fvt/storage/storage_test.go:
```diff
@@ -0,0 +1,189 @@
+			// recreate the storage-config secret without the PVCs
+			FVTClientInstance.CreateStorageConfigSecret(StorageConfigDataMinio)
+
+			// after applying configmap, the runtime pod(s) restart, wait for stability
```
With `scaleToZero`, there shouldn't be any "rollout" of pods since there are no ISVCs, right?

ckadner replied:
I will remove the scaleToZero part as it was an attempt to work around the unstable deployment problem

ckadner replied:
I removed the scaleToZero precondition since users would/should not be required to do this in order to get an _allow-any-PVC_ predictor deployed

tjohnson31415 commented on fvt/storage/storage_test.go:
```diff
@@ -0,0 +1,189 @@
+			FVTClientInstance.PrintPods()
+
+			// after mounting the new PVC the runtime pod(s) restart again, but the ISVC
+			// if not scaleToZero, it could have landed on the previous runtime pod will
```
These two lines seem a bit jumbled. `scaleToZero` is enabled in the applied configmap. In that case, can the ISVC still land on the previous runtime pod?

ckadner replied:
sorry, I missed to clean this up. another artifact of the 30 plus iterations/commits trying to get the FVTs to run reliably

tjohnson31415 commented on fvt/testdata/isvcs/isvc-pvc-3.yaml:
```diff
@@ -0,0 +1,14 @@
+metadata:
+  name: isvc-pvc3
+  labels:
+    name: isvc-pvc3
```
NIT: Can the label here be removed?

ckadner replied:
I used it to fetch the predictor for debugging purposes since the actual isvc name is different every run

ckadner replied:
removed now

tjohnson31415 reviewed (commented):
Overall the FVT changes are looking good to me! Just some NITs and questions.

ckadner commented on apis/serving/v1alpha1/predictor_types.go:
```diff
@@ -4,7 +4,7 @@
 // You may obtain a copy of the License at
 //
-//     http://www.apache.org/licenses/LICENSE-2.0
+//	http://www.apache.org/licenses/LICENSE-2.0
```
this was done by golang-lint, running `make fmt` (outside the dev container)

njhill replied:
@ckadner I'm curious why this would have just changed though... `make fmt` is run on all builds I thought?

ckadner replied:
It only happens when I run `make fmt` locally on my Mac using Go 1.19

When I run `make fmt` inside the dev container (with Go 1.18), or `make run fmt` from my native Mac terminal, then the spaces won't get replaced. 

I found this comment on golangci/golangci-lint#3063 but not sure it's the same thing

> Due to a change of go fmt inside go1.19.
> We have to enforce the directive style.
>
> So the option allow-leading-space has been dropped.
>
> The syntax of directives is: [a-zA-Z]+:[a-zA-Z].*.
> The spaces around : and after the // must be removed.

tjohnson31415 replied:
Ah, that is interesting. The [Go 1.19 Release Notes](https://tip.golang.org/doc/go1.19#go-doc) mention a change to `gofmt` that causes it to reformat godoc comments. Because the copyright headers are placed directly above the `package` declaration, `gofmt` for 1.19 is interpreting them as godocs and formatting them. (I verified that adding a blank line above the package directive prevents the change).

The effect that the change to `gofmt` has on our code will have to be addressed, but I'd prefer that it happens in a separate PR.

ckadner replied:
agreed, I undid those changes for now

ckadner replied:
I created issue #347 to track this

njhill commented:
Huge thanks @ckadner for all of the great work on this. And big thanks to @tjohnson31415 too for the thorough review!

ckadner commented on Makefile:
```diff
@@ -47,7 +47,7 @@ test:
 # Run fvt tests. This requires an etcd, kubernetes connection, and model serving installation. Ginkgo CLI is used to run them in parallel
 fvt:
-	ginkgo -v -p -progress --fail-fast fvt/predictor fvt/scaleToZero --timeout=40m
+	ginkgo -v -progress --fail-fast --flake-attempts=3 fvt/predictor fvt/scaleToZero fvt/storage --timeout=40m
```
In some circumstances the parallel runs caused problems when one test makes changes to the ModelMesh config while another does not need or expect that change (updating `storage-config secret`, changing `allowAnyPVC` flag)

ckadner replied:
oh, and the broken port-forwards when multiple one process disconnects from model serving while another process is just doing the inferencing

ckadner replied:
I guess we could split the predictor test suite into a part that could run in parallel processes and a part that needs sequential execution to make sure the differing configuration states don't cause side effects

tjohnson31415 replied:
Yeah, there are good reasons to limit the parallelization in certain test cases. [Ginkgo's `Serial` and `Ordered` decorators](https://onsi.github.io/ginkgo/#serial-specs) can help with that within a suite (that is what the Predictor suite uses for the TLS tests).

I'll leave it up to you if parallelization is worth it. As it stands, this PR looks to increase the FVT runtime by 50% from ~20mins to ~30mins.

ckadner replied:
**Ginkgo ran 3 suites in 10min 18sec**

```
Running Suite: Predictor suite - /fvt/predictor
=======================================================================================
Will run 116 of 117 specs
Running in parallel across 2 processes
...
------------------------------
Ran 116 of 117 Specs in 284.455 seconds
SUCCESS! -- 116 Passed | 0 Failed | 1 Pending | 0 Skipped
```
```
Running Suite: Scale to zero suite - /fvt/scaleToZero
=======================================================================================
Random Seed: 1679131805
Will run 5 of 5 specs
Running in parallel across 2 processes
...
------------------------------
Ran 5 of 5 Specs in 75.865 seconds
SUCCESS! -- 5 Passed | 0 Failed | 0 Pending | 0 Skipped
```
```
Running Suite: Storage Suite - /fvt/storage
=======================================================================================
Random Seed: 1679131805
Will run 9 of 9 specs
Running in parallel across 2 processes
...
------------------------------
Ran 9 of 9 Specs in 122.449 seconds
SUCCESS! -- 9 Passed | 0 Failed | 0 Pending | 0 Skipped

Ginkgo ran 3 suites in 10m18.221088608s
Test Suite Passed
```

ckadner replied:
https://github.com/ckadner/modelmesh-serving/blob/b3847c803c76e7578a670d94151cab9fa3eb7891/Makefile#L49-L50

ckadner replied:
> Yeah, there are good reasons to limit the parallelization in certain test cases. [Ginkgo's `Serial` and `Ordered` decorators](https://onsi.github.io/ginkgo/#serial-specs) can help with that within a suite (that is what the Predictor suite uses for the TLS tests).

Seems to not always work that way though, like in this past run, where the TLS test container started to run while the "New Format isvc" test was running as well:

```
 Summarizing 2 Failures:
  [FAIL] Inference service test New Format isvc [It] should successfully load a model
  /home/runner/work/modelmesh-serving/modelmesh-serving/fvt/helpers.go:296
  [INTERRUPTED] Predictor TLS XGBoost inference [It] should successfully run an inference with basic TLS
  /home/runner/work/modelmesh-serving/modelmesh-serving/fvt/predictor/predictor_test.go:883

Ran 111 of 117 Specs in 261.927 seconds
FAIL! - Interrupted by Other Ginkgo Process -- 109 Passed | 2 Failed | 1 Pending | 5 Skipped
```

ckadner replied:
I moved the `isvc_test`s into the `predictor_test` suite to guarantee they won't run in parallel to the `Serial` TLS tests. Seems to work reliably now, at least for the past 8 rounds of FVT runs I observed, 3 of them on this PR

ckadner commented on docs/install/README.md:
```diff
@@ -8,7 +8,7 @@
 - **etcd** - ModelMesh Serving requires an [etcd](https://etcd.io/) server in order to coordinate internal state which can be either dedicated or shared. More on this later.
 
-- **S3-compatible object storage** - Before models can be deployed, a remote S3-compatible datastore is needed from which to pull the model data. This could be for example an [IBM Cloud Object Storage](https://www.ibm.com/cloud/object-storage) instance, or a locally running [MinIO](https://github.com/minio/minio) deployment. Note that this is not required to be in place prior to the initial insta ...
+- **S3-compatible object storage** - Before models can be deployed, a remote S3-compatible datastore is needed from which to pull the model data. This could be for example an [IBM Cloud Object Storage](https://www.ibm.com/cloud/object-storage) instance, or a locally running [MinIO](https://github.com/minio/minio) deployment. Note that this is not required to be in place prior to the initial insta ...
```
How about this?

- **model storage** - The model files to be served by ModelMesh have to be stored either on S3-compatible object storage or on Kubernetes Persistent Volumes. For more information about supported storage options take a look at our [Setup Storage page](/docs/predictors/setup-storage.md).

tjohnson31415 replied:
That sounds better to me. Hopefully more storage providers will be added in the future (the puller technically supports HTTP, but we don't have documented usage of it).

ckadner replied:
The "final" wording looks like this now:

https://github.com/kserve/modelmesh-serving/blame/b3847c803c76e7578a670d94151cab9fa3eb7891/docs/install/install-script.md#L12

ckadner replied:
And I created issue #348 to document HTTP support in the future

ckadner replied:
Updated in this commit [9765cb1](https://github.com/kserve/modelmesh-serving/pull/340/commits/9765cb18d5871ad28f06c597a4bdad23e88040d6)

ckadner commented on fvt/fvtclient.go:
```diff
@@ -57,8 +57,8 @@ import (
-const predictorTimeout = time.Second * 120
-const timeForStatusToStabilize = time.Second * 5
+const PredictorTimeout = time.Second * 120        // abs time to wait for predictor to become ready
+const timeForStatusToStabilize = time.Second * 30 // max time in between watcher events
```
unfortunately, yes. You can take a look at the last couple of commits where I increase the `timeForStatusToStabilize` twice to address the FVT failures due to timeout exceeded. The only way to reduce the time penalty on all tests would be to isolate the test cases that need more time and parameterize the `WaitForRuntimeDeploymentsToBeStable` function

ckadner replied:
I introduced a `timeToStabilize` paramter to allow the slower tests to take more time while maintaining speedier stabilization times for the remaining ones

ckadner replied:
I guess I still have to recalibrate that :-\

https://github.com/kserve/modelmesh-serving/actions/runs/4453810628/jobs/7822588057?pr=340#step:13:140

ckadner replied:
I tweaked it a bit more. the default `timeForStatusToStabilize` = 5 seconds again

ckadner replied:
I reinstated the original timeouts:

https://github.com/kserve/modelmesh-serving/blob/b3847c803c76e7578a670d94151cab9fa3eb7891/fvt/fvtclient.go#L60-L61

ckadner commented on fvt/fvtclient.go:
```diff
@@ -494,21 +536,46 @@ func (fvt *FVTClient) RunTorchserveInference(req *torchserveapi.PredictionsReque
 	if fvt.grpcPortForward == nil {
-		podName := fvt.GetRandomReadyRuntimePodNameFromEndpoints()
+		//podName := fvt.GetRandomRuntimePodNameFromEndpoints()
+		podName := fvt.GetRandomReadyRuntimePod()
```
I had to look for a while to see why I did not always get "ready" pods until I found that particular method to check in the KNative repo

ckadner replied:
```go
	// make sure to only return pods that are 'Ready'
        // https://github.com/knative-sandbox/eventing-kafka-broker/pull/2523/files#diff-a9f3c3b
        for _, c := range p.Status.Conditions {
	        if c.Type == corev1.PodReady && c.Status == corev1.ConditionTrue {
		        pods.Items = append(pods.Items, p)
	        }
        }
```

ckadner commented on fvt/predictor/predictor_test.go:
```diff
@@ -878,6 +878,8 @@ var _ = Describe("Predictor", func() {
 		AfterAll(func() {
 			FVTClientInstance.SetDefaultUserConfigMap()
+			By("Waiting for a stable deploy state after changing the user config")
+			WaitForStableActiveDeployState()
```
I did not change the predictor_tests, but since those tests change the user config that is expected by other tests, I assume it was done to get back into the default state

ckadner replied:
I only added the wait since I struggled to get stable deploys

ckadner commented:
Thank you @tjohnson31415 and @njhill -- I will make the code changes you suggested and tag you for a re-review once the checks completed

ckadner commented on fvt/testdata/isvcs/isvc-pvc-4.yaml:
```diff
@@ -0,0 +1,14 @@
+metadata:
+  name: isvc-pvc4
+  labels:
+    name: isvc-pvc4
```
yup

ckadner commented on controllers/servingruntime_controller.go:
```diff
@@ -345,9 +358,50 @@ func (r *ServingRuntimeReconciler) getPVCs(ctx context.Context, req ctrl.Request
+	for claimName := range storageConfigPVCsMap {
+		r.Log.V(2).Info("Add PVC from storage-config to runtime",
+			"claimName", claimName,
+			"runtime", rt.BuiltInAdapter.ServerType)
```
okay

ckadner commented:
@tjohnson31415 -- I think I addressed all of your comments and created new issues to track outstanding followup work items. The review comments which were straight forward to address I collapsed as "Resolved" (Outdated).

I left some conversations as unresolved. Those you can mark as resolved if you are okay with my approach. If not let me know and I take another stab at them.  Thanks again for your thorough review!!!

**New issues:**
- #347 
- #348

ckadner commented:
@tjohnson31415 -- I think I addressed all of your comments and created new issues to track outstanding followup work items. The review comments which were straight forward to address I collapsed as "Resolved" (Outdated).

I left some conversations as unresolved. Those you can mark as resolved if you are okay with my approach. If not let me know and I take another stab at them.  Thanks again for your thorough review!!!

- [ ] [license header tab replacements with `gofmt` on Go 1.19
](https://github.com/kserve/modelmesh-serving/pull/340#discussion_r1137433326)
- [ ] [run fvt in parallel processes](https://github.com/kserve/modelmesh-serving/pull/340#discussion_r1137444863)
- [ ] [model storage README text change](https://github.com/kserve/modelmesh-serving/pull/340#discussion_r1140612265)
- [ ] [`timeForStatusToStabilize` = 5s](https://github.com/kserve/modelmesh-serving/pull/340#discussion_r1140621019)
- [ ] [random ready runtime pod](https://github.com/kserve/modelmesh-serving/pull/340#discussion_r1140628693)
- [ ] [my mistaken comment about allowAnyPVC needs rest-proxy enabled](https://github.com/kserve/modelmesh-serving/pull/340#discussion_r1140673018)
- [ ] [`scaleToZero` precondition in `allowAnyPVC` test case](https://github.com/kserve/modelmesh-serving/pull/340#discussion_r1140676682)
- [ ] [`WaitForStableActiveDeployState` after `SetDefaultUserConfigMap` in predictor tests](https://github.com/kserve/modelmesh-serving/pull/340#discussion_r1140685437)


**New issues:**
- #347 
- #348

ckadner commented:
@tjohnson31415 @njhill -- I [updated the "Model storage" text](https://github.com/kserve/modelmesh-serving/pull/340/commits/9765cb18d5871ad28f06c597a4bdad23e88040d6) bits and [removed the random ready endpoints code](https://github.com/kserve/modelmesh-serving/pull/340/commits/61982403a8cb20010a27d4e47e0041e19087dbf5).

I truly hope this is the last time I am asking you to review this PR :-)

ckadner commented on controllers/modelmesh/runtime.go:
```diff
@@ -69,6 +69,7 @@ func (m *Deployment) addVolumesToDeployment(deployment *appsv1.Deployment) error
 	}
 
 	// need to mount storage config for built-in adapters and the scenarios where StorageHelper is not disabled
+	// TODO: what about allowAnyPVC only use case without MinIO
```
I will create an issue to follow up

tjohnson31415 reviewed (approved):
More tests, less bugs, and better code. LGTM!
Thanks for going through the final iterations to keep the FVT duration down!

ckadner commented:
Thanks @tjohnson31415 for your review and all the valuable pointers!

@njhill I simplified the first PR comment to get a more concise commit message, but feel free to edit it before adding you `/lgmt` + `/approve`

njhill reviewed (approved):
Thanks again @ckadner @tjohnson31415

njhill commented:
/lgtm
